ReactiveCocoa/Swift/SignalProducer.swift

```diff
@@ -1674,172 +1672,188 @@ extension SignalProducerProtocol {
-	/// if necessary.
-	mutating func add(_ value: Value, upTo capacity: Int) {
-		precondition(capacity >= 0)
+	/// Initialize the replay state.
+	///
+	/// - parameters:
+	///   - capacity: The maximum amount of values which can be cached by the
+	///               replay state.
+	init(upTo capacity: Int) {
+		self.capacity = capacity
+	}
 
-		for buffer in replayBuffers {
-			buffer.values.append(value)
+	/// Attempt to observe the replay state.
+	///
+	/// - warning: Repeatedly observing the replay state with the same observer
+	///            should be avoided.
+	///
+	/// - parameters:
+	///   - observer: The observer to be registered.
+	///
+	/// - throws:
+	///   `ReplayError.pending(values:)` if there are still one or more values
+	///   pending for being replayed by the observer.
+	///
+	/// - returns:
+	///   The removal token of the observer, or `nil` if the replay state has
+	///   already terminated.
+	mutating func observe(_ observer: Signal<Value, Error>.Observer) throws -> RemovalToken? {
+		// The only use case is in `replayLazily` which has always a unique observer
+		// created for each produced signal. So we can use the ObjectIdentifier to
+		// track them directly.
+		let id = ObjectIdentifier(observer)
+
+		if let buffer = replayBuffers[id] {
+			if buffer.values.isEmpty {
+				replayBuffers.removeValue(forKey: id)
+				_ = terminationEvent.map(observer.action)
+				return observers?.insert(observer)
+			} else {
+				defer { buffer.values.removeAll() }
+				throw ReplayError<Value>.pending(values: buffer.values)
+			}
+		} else {
+			if values.isEmpty {
+				_ = terminationEvent.map(observer.action)
+				return observers?.insert(observer)
+			} else {
+				replayBuffers[id] = ReplayBuffer<Value>()
+				throw ReplayError<Value>.pending(values: values)
+			}
 		}
+	}
 
-		if capacity == 0 {
-			values = []
-			return
-		}
+	/// Enqueue the supplied event to the replay state.
+	///
+	/// - parameter:
+	///   - event: The event to be cached.
+	mutating func enqueue(_ event: Event<Value, Error>) {
+		if let value = event.value {
+			for buffer in replayBuffers.values {
+				buffer.values.append(value)
+			}
 
-		if capacity == 1 {
-			values = [ value ]
-			return
-		}
+			if capacity == 0 {
+				// `state.values` cannot be non-empty with a capacity of zero.
+				return
+			}
+
+			if capacity == 1 {
+				values = [value]
+				return
+			}
 
-		values.append(value)
+			values.append(value)
 
-		let overflow = values.count - capacity
-		if overflow > 0 {
-			values.removeSubrange(0..<overflow)
+			let overflow = values.count - capacity
+			if overflow > 0 {
+				values.removeSubrange(0 ..< overflow)
+			}
+		} else {
+			// Disconnect all observers and prevent future
+			// attachments.
+			terminationEvent = event
+			observers = nil
 		}
 	}
+
+	/// Broadcast the event to all observers.
+	///
+	/// - parameter:
+	///   - event: The event to be broadcast.
+	func broadcast(_ event: Event<Value, Error>) {
```

Comment:
I don't love having a side-effecting method on a value struct like this. I think it'd be better to move the code inline.